src/client.rs

```diff
@@ -97,47 +101,94 @@ impl ActivityPubClient {
+            .actor(
+                ActorBuilder::new(String::from("Actor")).url(
+                    actor_uri
+                        .parse::<http::Uri>()
+                        .map_err(|_e| Error::InvalidUri { uri: actor_uri })?,
+                ),
+            )
+            .to(vec![actor.id.as_ref().expect("actor has no id").clone()])
+            .object(ObjectBuilder::new().id(id))
+            .id(message_id_uri
+                .parse::<http::Uri>()
+                .map_err(|_e| Error::InvalidUri {
+                    uri: message_id_uri,
+                })?)
+            .build();
+
+        self.json_post(actor.inbox.expect("actor has no inbox"), message)
+            .await?;
 
         Ok(())
     }
 
     pub async fn unfollow_actor(&self, actor_uri: &str) -> Result<()> {
-        let Actor { id, inbox, .. } = self.get_actor(actor_uri).await?;
         let base = &self.base;
-        info!(%id, %inbox, "sending unfollow request to inbox");
+        let actor: Actor = self.get_actor(actor_uri).await?;
+        info!(
+            "sending unfollow request to inbox: {:?} {:?}",
+            actor.id, actor.inbox
+        );
 
         let object_id = Uuid::new_v4();
         let message_id = Uuid::new_v4();
-        let message = Activity {
-            context: Context::default(),
-            ty: ActivityType::Undo,
-            to: vec![actor_uri.to_owned()],
-            object: IdOrObject::Object {
-                ty: ActivityType::Follow,
-                object: actor_uri.to_owned(),
-                actor: actor_uri.to_owned(),
-                id: format!("https://{base}/activities/{object_id}"),
-            },
-            id: format!("https://{base}/activities/{message_id}"),
-            actor: format!("https://{base}/actor)"),
-        };
-
-        self.json_post(inbox, message).await?;
+        let activity_id = format!("https://{base}/activities/{message_id}");
+        let activity_id_uri = activity_id
+            .parse::<http::Uri>()
+            .map_err(|_e| Error::InvalidUri { uri: activity_id })?;
+        let object_id = format!("https://{base}/activities/{object_id}");
+        let object_id_uri = object_id
+            .parse::<http::Uri>()
+            .map_err(|_e| Error::InvalidUri { uri: object_id })?;
+
+        let message = ActivityBuilder::new(String::from("Undo"), String::from("Unfollow actor"))
+            .actor(
+                ActorBuilder::new(String::from("Actor")).url(
+                    format!("https://{base}/actor")
+                        .parse::<http::Uri>()
+                        .map_err(|_e| Error::InvalidUri {
+                            uri: format!("https://{base}/actor"),
+                        })?,
+                ),
+            )
+            .to(vec![actor_uri.to_owned()])
+            .object(
+                // FIXME: object does not have a property "actor":
```

Comment:
The message formats that I was using before were taken from the relay implementation we were referencing that we've now found out was incorrect. We likely need to review / rewrite the payloads we're currently sending I'm afraid.

Reply:
ok. i'll add this to the FIXME here and we can get back to it :)